synapse/types.py

```diff
@@ -400,6 +400,37 @@ def f2(m: Match[bytes]) -> bytes:
     return username.decode("ascii")
 
 
+MU = TypeVar("MU", bound="MXCUri")
+
+
+@attr.s(frozen=True, slots=True, auto_attribs=True, repr=False)
+class MXCUri:
+    """Represents a URI that points to a media resource in matrix.
+
+    MXC's take the form 'mxc://server_name/media_id'.
+    """
+
+    server_name: str
+    media_id: str
+
+    @classmethod
+    def from_str(cls: Type[MU], mxc_uri_str: str) -> MU:
+        try:
+            server_name, media_id = mxc_uri_str.split("/")[-2:]
+        except IndexError:
+            raise SynapseError(
+                400, f"Error processing malformed MXC uri: {mxc_uri_str}"
+            )
+
+        return cls(server_name, media_id)
+
+    def to_string(self) -> str:
+        """Convert an MXCUri object to a str."""
+        return f"mxc://{self.server_name}/{self.media_id}"
+
+    __repr__ = to_string
```

Comment:
This is a bit naughty. To quote https://docs.python.org/3/reference/datamodel.html?highlight=__repr__#object.__repr__ :+1: 

> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment).

I would use the attrs-generated `__repr__` and write `__str__ = to_string` instead.

Reply:
Ah! I cribbed this behaviour from `DomainSpecificString`, but the documentation has cleared up why that would be a bad idea. I suppose `DomainSpecificString` could get a pass as it's supposed to simply be a str?

Reply:
I'd change DomainSpecificString too, were it not for the fear of breaking everything that already uses it.

Reply:
I'm going to leave that alone for now then.

Reply:
> I suppose DomainSpecificString could get a pass as it's supposed to simply be a str?

Not sure I'd agree on that: it's a parsed version of some data that happens to be encoded as a string.

Even then: for `x: str`, `repr(x) != str(x)`; the repr includes quotation marks.

Reply:
Good point, I could see it being hard to debug in some particularly complex usage of `DomainSpecificString`.